crates/iceberg-catalog/src/service/storage/gcs/mod.rs

```diff
@@ -0,0 +1,373 @@
+            .as_ref()
+            .map(|s| s.split('/').map(std::borrow::ToOwned::to_owned).collect())
+            .unwrap_or_default();
+        Location::from_str(&format!("gs://{}/", self.bucket))
+            .map(|mut l| {
+                l.extend(prefix.iter());
+                l
+            })
+            .map_err(|e| ValidationError::InvalidLocation {
+                reason: "Invalid GCS location.".to_string(),
+                location: format!("gs://{}/", self.bucket),
+                source: Some(e.into()),
+                storage_type: StorageType::Gcs,
+            })
+    }
+
+    /// Generate the table configuration for GCS.
+    pub(crate) async fn generate_table_config(
+        &self,
+        _: &DataAccess,
+        cred: Option<&GcsCredential>,
+        table_location: &Location,
+        storage_permissions: StoragePermissions,
+    ) -> Result<TableProperties, TableConfigError> {
+        let mut config = TableProperties::default();
+        if let Some(GcsCredential::ServiceAccountKey { key }) = cred {
+            let token = sts::downscope(
+                key,
+                &self.bucket,
+                table_location.clone(),
+                storage_permissions,
+            )
+            .await?;
+
+            config.insert(&gcs::Token(token.access_token));
+            config.insert(&gcs::ProjectId(key.project_id.clone()));
+
+            if let Some(expiry) = token.expires_in {
+                config.insert(&gcs::TokenExpiresAt(
+                    (std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)
+                        .unwrap()
+                        .as_secs()
+                        + expiry as u64)
+                        .to_string(),
+                ));
+            }
+        }
+
+        Ok(config)
+    }
+
+    fn normalize_key_prefix(&mut self) -> Result<(), ValidationError> {
+        if let Some(key_prefix) = self.key_prefix.as_mut() {
+            *key_prefix = key_prefix.trim_matches('/').to_string();
+            if key_prefix.starts_with(".well-known/acme-challenge/") {
+                return Err(ValidationError::InvalidProfile {
+                    source: None,
+                    reason: "Storage Profile `key_prefix` cannot start with `.well-known/acme-challenge/`.".to_string(),
+                    entity: "key_prefix".to_string(),
+                });
+            }
+        }
+
+        if let Some(key_prefix) = self.key_prefix.as_ref() {
+            if key_prefix.is_empty() {
+                self.key_prefix = None;
+            }
+        }
+
+        // GCS supports a max of 1024 chars and we need some buffer for tables.
+        if let Some(key_prefix) = self.key_prefix.as_ref() {
+            if key_prefix.len() > 896 {
+                return Err(ValidationError::InvalidProfile {
+                    source: None,
+                    reason: "Storage Profile `key_prefix` must be less than 1024 characters."
```

Comment:
Less than 896 chars?

Reply:
copy paste error from s3, also fixing it over there 😄